sdk/metric/pipeline.go

```diff
@@ -134,3 +138,111 @@ func (p *pipeline) produce(ctx context.Context) (metricdata.ResourceMetrics, err
+type pipelineRegistry[N int64 | float64] struct {
+	views     map[Reader][]view.View
+	pipelines map[Reader]*pipeline
+}
+
+func newPipelineRegistries(views map[Reader][]view.View) (*pipelineRegistry[int64], *pipelineRegistry[float64]) {
+	pipelines := map[Reader]*pipeline{}
+	for rdr := range views {
+		pipe := &pipeline{}
+		rdr.register(pipe)
+		pipelines[rdr] = pipe
+	}
+	return &pipelineRegistry[int64]{
+			views:     views,
+			pipelines: pipelines,
+		}, &pipelineRegistry[float64]{
+			views:     views,
+			pipelines: pipelines,
+		}
+}
+
+// createAggregators will create all backing aggregators for an instrument.
+// It will return an error if an instrument is registered more than once.
+// Note: There may be returned aggregators with an error.
+func (reg *pipelineRegistry[N]) createAggregators(inst view.Instrument, instUnit unit.Unit) ([]internal.Aggregator[N], error) {
+	var aggs []internal.Aggregator[N]
+
+	errs := &multierror{}
+	for rdr, views := range reg.views {
+		pipe := reg.pipelines[rdr]
+		rdrAggs := createAggregators[N](rdr, views, inst)
+		for inst, agg := range rdrAggs {
+			err := pipe.addAggregator(inst.Scope, inst.Name, inst.Description, instUnit, agg)
+			if err != nil {
+				errs.append(err)
+			}
+			aggs = append(aggs, agg)
+		}
+	}
+	return aggs, errs.errorOrNil()
+}
+
+func (reg *pipelineRegistry[N]) registerCallback(fn func(context.Context)) {
+	for _, pipe := range reg.pipelines {
+		pipe.addCallback(fn)
+	}
+}
+
+type multierror struct {
+	errors []string
+}
+
+func (m *multierror) errorOrNil() error {
+	if len(m.errors) == 0 {
+		return nil
+	}
+	return fmt.Errorf(strings.Join(m.errors, "; "))
+}
+func (m *multierror) append(err error) {
+	m.errors = append(m.errors, err.Error())
+}
+
+func createAggregators[N int64 | float64](rdr Reader, views []view.View, inst view.Instrument) map[view.Instrument]internal.Aggregator[N] {
+	aggs := map[view.Instrument]internal.Aggregator[N]{}
+	for _, v := range views {
+		inst, match := v.TransformInstrument(inst)
+		if inst.Aggregation == nil {
+			inst.Aggregation = rdr.aggregation(inst.Kind)
+		}
+		if match {
+			if _, ok := aggs[inst]; ok {
+				continue
+			}
+			agg := createAggregator[N](inst.Aggregation, rdr.temporality(inst.Kind))
+			if agg != nil {
+				// TODO (#3011): If filtering is done at the instrument level add here.
+				// This is where the aggregator and the view are both in scope.
+				aggs[inst] = agg
+			}
+		}
+	}
+	return aggs
+}
+
+// createAggregator takes the config (Aggregation and Temporality) and produces a memory backed Aggregator.
+// TODO (#3011): If filterting is done by the Aggregator it should be passed here.
+func createAggregator[N int64 | float64](agg aggregation.Aggregation, temporality metricdata.Temporality) internal.Aggregator[N] {
+	switch agg := agg.(type) {
+	case aggregation.Drop:
+		return nil
+	case aggregation.LastValue:
+		return internal.NewLastValue[N]()
+	case aggregation.Sum:
```

Comment:
This will need to have context about what instrument this is to provide the temporality of the sum.

Reply:
Yes, I was only writing to the API that has landed.  The same can be said for the histogram.

Reply:
I guess this will fail to compile when it merges with #3000. Therefore, we will have a signal to follow when this needs to be updated and no need to track the issue. :+1: